### PR TITLE
[Wrangler] Added sub commands in the `deployments`

### PR DESCRIPTION
**Pull request by JacobMGEvans**
`view <deployment-id>` will get the details of a deployment, including versioned script, bindings, and usage model information.
This information can be used to help debug bad deployments or get insights on changes between deployments.

`rollback <deployment-id>` will rollback to a specific deployment in the runtime. This will be useful in situations like recovering from a bad
deployment quickly while resolving issues. This rollback only changes the code in the runtime and doesn't affect any code or configurations
in a developer's local setup.

example of `view <deployment-id>` output:

```ts
	{
    Tag: '',
    Number: 0,
    'Metadata.author_id': 'Picard-Gamma-6-0-7-3',
    'Metadata.author_email': 'picard@vinyard.com',
    'Metadata.source': 'wrangler',
    'Metadata.created_on': '2021-01-01T00:00:00.000000Z',
    'Metadata.modified_on': '2021-01-01T00:00:00.000000Z',
    'resources.script': {
      etag: 'mock-e-tag',
      handlers: [ 'fetch' ],
      last_deployed_from: 'wrangler'
    },
    'resources.bindings': []
  }



  export default {
    async fetch(request) {
      return new Response('Hello World from Deployment 1701-E');
    },
  };
```
- [x] Docs PR https://github.com/cloudflare/cloudflare-docs/pull/7492
- [x] Tests
- [x] Changeset

Reviewer has performed the following, where applicable:

- [ ] Checked for inclusion of relevant tests
- [ ] Checked for inclusion of a relevant changeset
- [ ] Checked for creation of associated docs updates
- [ ] Manually pulled down the changes and spot-tested

Ticket: Internal

### CODE REVIEW

File: packages/wrangler/src/__tests__/deployments.test.ts [hunk 1]
```diff
@@ -102,4 +110,116 @@ describe("deployments", () => {
...
+			  \\"handlers\\": [
+			    \\"interstellar_communication\\",
+			    \\"hyperspace_navigation\\"
+			  ],
+			  \\"last_deployed_from\\": \\"spaceport_alpha\\",
+			  \\"usage_model\\": \\"intergalactic\\",
+			  \\"script\\": \\"addEventListener('interstellar_communication', event =>/n/t/t/t/t/t/t/t{ event.respondWith(transmit(event.request)) }/n/t/t/t/t/t/t/t)\\",
+			  \\"size\\": \\"1 light-year\\"
+			}"
+		`);
```
Comment (JacobMGEvans) [inline packages/wrangler/src/__tests__/deployments.test.ts:221]: This is a placeholder until the message copy is determined for the success response. 
The prompt is also yet to be implemented.

File: packages/wrangler/src/__tests__/helpers/msw/handlers/deployments.ts [hunk 2]
```diff
@@ -53,17 +53,68 @@ export const mswSuccessDeployments = [
...
+			let scriptContent = "";
+			if (req.url.searchParams.get("deployment") === "1701-E") {
+				scriptContent = `
+			export default {
+				async fetch(request) {
+					return new Response('Hello World from Deployment 1701-E');
+				},
+			};`;
+			} else {
+				scriptContent = "DIDN'T GET SCRIPT CONTENT";
```
Comment (JacobMGEvans) [inline packages/wrangler/src/__tests__/helpers/msw/handlers/deployments.ts:115]: note to self: Create a test for script retrieval failure.

**Review by rozenmd — COMMENTED** [review]
Now that `wrangler deployments` is starting to actually do stuff, (in a separate PR), can we move the handler and options into its own file or folder, like `dev`, `delete`, `publish`, `r2`, `secret`, etc?

File: packages/wrangler/src/index.ts [hunk 3]
```diff
@@ -588,17 +588,66 @@ export function createCLIParser(argv: string[]) {
...
+							demandOption: true,
+						}),
+					async (rollbackYargs) => {
+						const firstHash = rollbackYargs.deploymentId.substring(
+							0,
+							rollbackYargs.deploymentId.indexOf("-")
+						);
+						if (
+							!(await confirm(
+								`This deployment ${firstHash} will immediately replace the current deployment and become the active deployment across all your deployed routes and domains. However, your local development environment will not be affected by this rollback.`
```
Comment (lrapoport-cf) [inline packages/wrangler/src/index.ts:607]: @JacobMGEvans  following our discussion earlier, let's add the same note here as in the dash confirmation modal. so to the end of this please add:

"Note: Rolling back to a previous deployment will not rollback any of the bound resources (Durable Object, R2, KV, etc.)."

File: packages/wrangler/src/publish/publish.ts [hunk 4]
```diff
@@ -828,19 +827,7 @@ See https://developers.cloudflare.com/workers/platform/compatibility-dates for m
...
-
-		logger.log("Current Deployment ID:", deploymentsList.latest.id);
-	} catch (e) {
-		if ((e as { code: number }).code === 10023) {
-			// TODO: remove this try/catch once deployments is completely rolled out
-		} else {
-			throw e;
-		}
-	}
+	logger.log("Current Deployment ID:", deploymentId);
```
Comment (jspspike) [inline packages/wrangler/src/publish/publish.ts:830]: This will cause the deployment ID to not hyphens, I'm not sure if that's unpolished/confused to users

Comment (jspspike) [reply packages/wrangler/src/publish/publish.ts:830]: Same [here](https://github.com/cloudflare/workers-sdk/pull/2607/files#diff-605d740517c44d6d11fe53bd4718c64bf8baa876c6f1af516e0aa5efcfe3c977R149)

File: packages/wrangler/src/__tests__/helpers/msw/handlers/deployments.ts (hunk 2 above)
Comment (JacobMGEvans) [reply packages/wrangler/src/__tests__/helpers/msw/handlers/deployments.ts:139]: And failed requests with these scenarios:
```
"errors": [
    {
      "code": 100146,
      "message": "workers.api.error.deployment_not_found"
    }
  ],
  "messages": []
```

```
"errors": [
    {
      "code": 100147,
      "message": "workers.api.error.deployment_too_old"
    }
  ],
  "messages": []
```

```
"errors": [
    {
      "code": 100140,
      "message": "Failed to rollback due to external resources being modified (durable-object:<tag>,r2:<name>)"
    }
  ],
  "messages": []
```

**Review by jspspike — APPROVED** [review]

File: packages/wrangler/index.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+		}
+		throw new Error("Not found!");
+	},
+	async scheduled(controller, env, ctx) {
+		const stmt = await env.DB.prepare(
+			"INSERT INTO test (id, value) VALUES (?, ?)"
+		);
+		await stmt.bind(1, "one").run();
+	},
+};
```
Comment (JacobMGEvans) [inline packages/wrangler/index.js:21]: Don't commit this @jspspike , its randomly generated by one of the tests, @rozenmd is this a D1 related test?

Comment (JacobMGEvans) [reply packages/wrangler/index.js]: I removed it.

**Review by petebacondarwin — APPROVED** [review]

File: packages/wrangler/src/__tests__/publish.test.ts [hunk 6]
```diff
@@ -7155,53 +7155,6 @@ export default{
 		});
 	});
 
-	it("should publish if the last deployed source check fails", async () => {
```
[the remark this answers is not in the document]
Comment (jspspike) [reply packages/wrangler/src/__tests__/publish.test.ts:7158]: This was removed since it tested if the `deployments` endpoint failed on publish. That API call was removed so this test is no longer necessary

File: packages/wrangler/src/__tests__/jest.setup.ts [hunk 7]
```diff
@@ -1,7 +1,11 @@
+import chalk from "chalk";
 import fetchMock from "jest-fetch-mock";
 import { MockWebSocket } from "./helpers/mock-web-socket";
 import { msw } from "./helpers/msw";
 
+//turn off chalk for tests
```
[the remark this answers is not in the document]
Comment (JacobMGEvans) [reply packages/wrangler/src/__tests__/jest.setup.ts:6]: Yes, has to do with different styling between OS's and as well as TTY and non-TTY

File: packages/wrangler/src/__tests__/publish.test.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (JacobMGEvans) [reply packages/wrangler/src/__tests__/publish.test.ts:7158]: Yeah, Josh streamlined the Publish (honestly should have been another PR) it now gets the deployment ID from the response on the PUT/POST rather than making additional calls to check the ID

File: .changeset/brown-wasps-fail.md [hunk 8]
```diff
@@ -0,0 +1,39 @@
+---
+"wrangler": minor
+---
+
+Added sub commands in the `deployments` command:
```
[the remark this answers is not in the document]
Comment (JacobMGEvans) [reply .changeset/brown-wasps-fail.md:5]: I will get better at this 😅

**Review by petebacondarwin — APPROVED** [review]
